File: 56-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        LinkedList<int[]> merged = new LinkedList<>();
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        for(int[] interval : intervals) {
            if(merged.isEmpty() || interval[0] > merged.getLast()[1]) {
                merged.add(interval);
            }
            else {
                merged.getLast()[1] = Math.max(merged.getLast()[1], interval[1]);
            }
        }
        
        return merged.toArray(new int[merged.size()][]);
    }
}

File: 56_Merge_Intervals.py
I sort the intervals by start time. Then I scan through them and compare each interval with the last merged one. 
If they overlap, I extend the end boundary. 
If not, I add a new interval. This runs in O(n log n) because of sorting and O(n) for the merge.

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # Sort intervals by start time
        intervals.sort(key=lambda x: x[0])

        merged = [intervals[0]]

        for start, end in intervals[1:]:
            last_start, last_end = merged[-1]

            # If overlapping: extend the last interval
            if start <= last_end:
                merged[-1][1] = max(last_end, end)
            else:
                merged.append([start, end])
        
        return merged
